File: API/0450_Tendency_Ema_Rsi/PY/tendency_ema_rsi_strategy.py
import clr

clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")

from StockSharp.Messages import DataType, CandleStates, Sides
from StockSharp.Algo.Indicators import RelativeStrengthIndex, ExponentialMovingAverage
from StockSharp.Algo.Strategies import Strategy
from datatype_extensions import *


class tendency_ema_rsi_strategy(Strategy):
    """EMA crossover with RSI filter and trend confirmation."""

    def __init__(self):
        super(tendency_ema_rsi_strategy, self).__init__()

        self._candle_type = self.Param("CandleType", tf(1)) \
            .SetDisplay("Candle Type", "Type of candles to use", "General")

        self._rsi_length = self.Param("RsiLength", 14) \
            .SetDisplay("RSI Length", "RSI calculation length", "RSI")

        self._ema_a_length = self.Param("EmaALength", 9) \
            .SetDisplay("EMA A Length", "Fast EMA", "Moving Averages")

        self._ema_b_length = self.Param("EmaBLength", 21) \
            .SetDisplay("EMA B Length", "Medium EMA", "Moving Averages")

        self._ema_c_length = self.Param("EmaCLength", 50) \
            .SetDisplay("EMA C Length", "Slow EMA", "Moving Averages")

        self._show_long = self.Param("ShowLong", True) \
            .SetDisplay("Long Entries", "Enable long entries", "Strategy")

        self._show_short = self.Param("ShowShort", False) \
            .SetDisplay("Short Entries", "Enable short entries", "Strategy")

        self._close_after_x = self.Param("CloseAfterXBars", False) \
            .SetDisplay("Close After X Bars", "Close position after X bars if in profit", "Strategy")

        self._x_bars = self.Param("XBars", 5) \
            .SetDisplay("X Bars", "Number of bars to hold", "Strategy")

        self._rsi = None
        self._ema_a = None
        self._ema_b = None
        self._ema_c = None

        self._prev_ema_a = 0
        self._prev_ema_b = 0
        self._cross_over = False
        self._cross_under = False
        self._bars_in_pos = 0
        self._entry_price = 0

    # region properties
    @property
    def candle_type(self):
        return self._candle_type.Value

    @property
    def rsi_length(self):
        return self._rsi_length.Value

    @property
    def ema_a_length(self):
        return self._ema_a_length.Value

    @property
    def ema_b_length(self):
        return self._ema_b_length.Value

    @property
    def ema_c_length(self):
        return self._ema_c_length.Value

    @property
    def show_long(self):
        return self._show_long.Value

    @property
    def show_short(self):
        return self._show_short.Value

    @property
    def close_after_x(self):
        return self._close_after_x.Value

    @property
    def x_bars(self):
        return self._x_bars.Value

    # endregion

    def GetWorkingSecurities(self):
        return [(self.Security, self.candle_type)]

    def OnStarted(self, time):
        super().OnStarted(time)

        self._rsi = RelativeStrengthIndex(Length=self.rsi_length)
        self._ema_a = ExponentialMovingAverage(Length=self.ema_a_length)
        self._ema_b = ExponentialMovingAverage(Length=self.ema_b_length)
        self._ema_c = ExponentialMovingAverage(Length=self.ema_c_length)

        sub = self.SubscribeCandles(self.candle_type)
        sub.Bind(self._rsi, self._ema_a, self._ema_b, self._ema_c, self.ProcessCandle).Start()

        area = self.CreateChartArea()
        if area is not None:
            self.DrawCandles(area, sub)
            self.DrawIndicator(area, self._ema_a)
            self.DrawIndicator(area, self._ema_b)
            self.DrawIndicator(area, self._ema_c)
            self.DrawOwnTrades(area)

    def ProcessCandle(self, candle, rsi_value, ema_a_val, ema_b_val, ema_c_val):
        if candle.State != CandleStates.Finished:
            return

        if not self._rsi.IsFormed or not self._ema_a.IsFormed or not self._ema_b.IsFormed or not self._ema_c.IsFormed:
            return

        price = candle.ClosePrice
        open_price = candle.OpenPrice

        if self._prev_ema_a and self._prev_ema_b:
            self._cross_over = self._prev_ema_a <= self._prev_ema_b and ema_a_val > ema_b_val
            self._cross_under = self._prev_ema_a >= self._prev_ema_b and ema_a_val < ema_b_val

        if self.Position != 0:
            self._bars_in_pos += 1
        else:
            self._bars_in_pos = 0
            self._entry_price = 0

        self.CheckEntry(candle, rsi_value, ema_a_val, ema_b_val, ema_c_val)
        self.CheckExit(candle, rsi_value, ema_a_val, ema_b_val, ema_c_val)

        self._prev_ema_a = ema_a_val
        self._prev_ema_b = ema_b_val

    def CheckEntry(self, candle, rsi_val, ema_a_val, ema_b_val, ema_c_val):
        price = candle.ClosePrice
        open_price = candle.OpenPrice

        if (self.show_long and self._cross_over and ema_a_val > ema_c_val and price > open_price and self.Position == 0):
            self._entry_price = open_price
            self.RegisterOrder(self.CreateOrder(Sides.Buy, price, self.Volume))

        if (self.show_short and self._cross_under and ema_a_val < ema_c_val and price < open_price and self.Position == 0):
            self._entry_price = open_price
            self.RegisterOrder(self.CreateOrder(Sides.Sell, price, self.Volume))

    def CheckExit(self, candle, rsi_val, ema_a_val, ema_b_val, ema_c_val):
        price = candle.ClosePrice

        if self.Position > 0 and rsi_val > 70:
            self.RegisterOrder(self.CreateOrder(Sides.Sell, price, abs(self.Position)))
            return

        if self.Position < 0 and rsi_val < 30:
            self.RegisterOrder(self.CreateOrder(Sides.Buy, price, abs(self.Position)))
            return

        if self.close_after_x and self._entry_price != 0 and self._bars_in_pos >= self.x_bars:
            if self.Position > 0 and price > self._entry_price:
                self.RegisterOrder(self.CreateOrder(Sides.Sell, price, abs(self.Position)))
            elif self.Position < 0 and price < self._entry_price:
                self.RegisterOrder(self.CreateOrder(Sides.Buy, price, abs(self.Position)))

    def CreateClone(self):
        """
        !! REQUIRED!! Creates a new instance of the strategy.
        """
        return tendency_ema_rsi_strategy()

File: API/0450_Tendency_Ema_Rsi/CS/TendencyEmaRsiStrategy.cs
namespace StockSharp.Samples.Strategies;

using System;
using System.Collections.Generic;

using Ecng.Common;

using StockSharp.Algo.Indicators;
using StockSharp.Algo.Strategies;
using StockSharp.BusinessEntities;
using StockSharp.Messages;

/// <summary>
/// Tendency EMA + RSI Strategy.
/// Uses EMA crossover with RSI and trend filter.
/// Buys when fast EMA crosses above medium EMA while above slow EMA.
/// Exits when RSI becomes overbought/oversold.
/// </summary>
public class TendencyEmaRsiStrategy : Strategy
{
	private readonly StrategyParam<DataType> _candleType;
	private readonly StrategyParam<int> _rsiLength;
	private readonly StrategyParam<int> _emaALength;
	private readonly StrategyParam<int> _emaBLength;
	private readonly StrategyParam<int> _emaCLength;
	private readonly StrategyParam<int> _cooldownBars;

	private RelativeStrengthIndex _rsi;
	private ExponentialMovingAverage _emaA;
	private ExponentialMovingAverage _emaB;
	private ExponentialMovingAverage _emaC;

	private decimal _prevEmaA;
	private decimal _prevEmaB;
	private int _cooldownRemaining;

	public DataType CandleType
	{
		get => _candleType.Value;
		set => _candleType.Value = value;
	}

	public int RsiLength
	{
		get => _rsiLength.Value;
		set => _rsiLength.Value = value;
	}

	public int EmaALength
	{
		get => _emaALength.Value;
		set => _emaALength.Value = value;
	}

	public int EmaBLength
	{
		get => _emaBLength.Value;
		set => _emaBLength.Value = value;
	}

	public int EmaCLength
	{
		get => _emaCLength.Value;
		set => _emaCLength.Value = value;
	}

	public int CooldownBars
	{
		get => _cooldownBars.Value;
		set => _cooldownBars.Value = value;
	}

	public TendencyEmaRsiStrategy()
	{
		_candleType = Param(nameof(CandleType), TimeSpan.FromMinutes(30).TimeFrame())
			.SetDisplay("Candle Type", "Type of candles to use", "General");

		_rsiLength = Param(nameof(RsiLength), 14)
			.SetGreaterThanZero()
			.SetDisplay("RSI Length", "RSI calculation length", "RSI");

		_emaALength = Param(nameof(EmaALength), 10)
			.SetGreaterThanZero()
			.SetDisplay("EMA A Length", "Fast EMA length", "Moving Averages");

		_emaBLength = Param(nameof(EmaBLength), 20)
			.SetGreaterThanZero()
			.SetDisplay("EMA B Length", "Medium EMA length", "Moving Averages");

		_emaCLength = Param(nameof(EmaCLength), 100)
			.SetGreaterThanZero()
			.SetDisplay("EMA C Length", "Slow/Trend EMA length", "Moving Averages");

		_cooldownBars = Param(nameof(CooldownBars), 10)
			.SetDisplay("Cooldown Bars", "Bars to wait between trades", "Risk");
	}

	/// <inheritdoc />
	public override IEnumerable<(Security sec, DataType dt)> GetWorkingSecurities()
		=> [(Security, CandleType)];

	/// <inheritdoc />
	protected override void OnReseted()
	{
		base.OnReseted();

		_rsi = null;
		_emaA = null;
		_emaB = null;
		_emaC = null;
		_prevEmaA = 0;
		_prevEmaB = 0;
		_cooldownRemaining = 0;
	}

	/// <inheritdoc />
	protected override void OnStarted2(DateTime time)
	{
		base.OnStarted2(time);

		_rsi = new RelativeStrengthIndex { Length = RsiLength };
		_emaA = new ExponentialMovingAverage { Length = EmaALength };
		_emaB = new ExponentialMovingAverage { Length = EmaBLength };
		_emaC = new ExponentialMovingAverage { Length = EmaCLength };

		var subscription = SubscribeCandles(CandleType);
		subscription
			.Bind(_rsi, _emaA, _emaB, _emaC, OnProcess)
			.Start();

		var area = CreateChartArea();
		if (area != null)
		{
			DrawCandles(area, subscription);
			DrawIndicator(area, _emaA);
			DrawIndicator(area, _emaB);
			DrawIndicator(area, _emaC);
			DrawOwnTrades(area);
		}
	}

	private void OnProcess(ICandleMessage candle, decimal rsiVal, decimal emaA, decimal emaB, decimal emaC)
	{
		if (candle.State != CandleStates.Finished)
			return;

		if (!_rsi.IsFormed || !_emaA.IsFormed || !_emaB.IsFormed || !_emaC.IsFormed)
		{
			_prevEmaA = emaA;
			_prevEmaB = emaB;
			return;
		}

		if (!IsFormedAndOnlineAndAllowTrading())
		{
			_prevEmaA = emaA;
			_prevEmaB = emaB;
			return;
		}

		if (_cooldownRemaining > 0)
		{
			_cooldownRemaining--;
			_prevEmaA = emaA;
			_prevEmaB = emaB;
			return;
		}

		if (_prevEmaA == 0 || _prevEmaB == 0)
		{
			_prevEmaA = emaA;
			_prevEmaB = emaB;
			return;
		}

		// EMA crossovers
		var emaCrossUp = emaA > emaB && _prevEmaA <= _prevEmaB;
		var emaCrossDown = emaA < emaB && _prevEmaA >= _prevEmaB;

		// Buy: EMA A crosses above EMA B + EMA A > EMA C (uptrend) + bullish candle
		if (emaCrossUp && emaA > emaC && candle.ClosePrice > candle.OpenPrice && Position <= 0)
		{
			if (Position < 0)
				BuyMarket(Math.Abs(Position));
			BuyMarket(Volume);
			_cooldownRemaining = CooldownBars;
		}
		// Sell: EMA A crosses below EMA B + EMA A < EMA C (downtrend) + bearish candle
		else if (emaCrossDown && emaA < emaC && candle.ClosePrice < candle.OpenPrice && Position >= 0)
		{
			if (Position > 0)
				SellMarket(Math.Abs(Position));
			SellMarket(Volume);
			_cooldownRemaining = CooldownBars;
		}
		// Exit long: RSI overbought
		else if (Position > 0 && rsiVal > 70)
		{
			SellMarket(Math.Abs(Position));
			_cooldownRemaining = CooldownBars;
		}
		// Exit short: RSI oversold
		else if (Position < 0 && rsiVal < 30)
		{
			BuyMarket(Math.Abs(Position));
			_cooldownRemaining = CooldownBars;
		}

		_prevEmaA = emaA;
		_prevEmaB = emaB;
	}
}
